Assets/HoloToolkit/Input/Scripts/Microphone/DictationInputManager.cs

```diff
@@ -0,0 +1,254 @@
+        /// </summary>
+        public static bool IsListening { get; private set; }
+
+        /// <summary>
+        /// String result of the current dictation.
+        /// </summary>
+        private static string dictationResult;
+
+        /// <summary>
+        /// Audio clip of the last dictation session.
+        /// </summary>
+        private static AudioClip dictationAudioClip;
+
+        private static DictationRecognizer dictationRecognizer;
+
+        private bool hasFailed;
+
+        #region Unity Methods
+
+        protected override void Awake()
+        {
+            base.Awake();
+
+            dictationResult = string.Empty;
+
+            dictationRecognizer = new DictationRecognizer();
+            dictationRecognizer.DictationHypothesis += DictationRecognizer_DictationHypothesis;
+            dictationRecognizer.DictationResult += DictationRecognizer_DictationResult;
+            dictationRecognizer.DictationComplete += DictationRecognizer_DictationComplete;
+            dictationRecognizer.DictationError += DictationRecognizer_DictationError;
+
+            // Query the maximum frequency of the default microphone.
+            int minSamplingRate; // Unsued.
+            Microphone.GetDeviceCaps(DeviceName, out minSamplingRate, out samplingRate);
+        }
+
+        private void LateUpdate()
+        {
+            if (IsListening && !Microphone.IsRecording(DeviceName) && dictationRecognizer.Status == SpeechSystemStatus.Running)
+            {
+                // If the microphone stops as a result of timing out, make sure to manually stop the dictation recognizer.
+                StartCoroutine(StopRecording());
+            }
+
+            if (!hasFailed && dictationRecognizer.Status == SpeechSystemStatus.Failed)
+            {
+                hasFailed = true;
+                InputManager.Instance.RaiseDictationError(Instance, 0, "Dictation recognizer has failed!");
+            }
+        }
+
+        protected override void OnDestroy()
+        {
+            dictationRecognizer.Dispose();
+
+            base.OnDestroy();
+        }
+
+        #endregion // Unity Methods
+
+        /// <summary>
+        /// Turns on the dictation recognizer and begins recording audio from the default microphone.
+        /// </summary>
+        /// <param name="initialSilenceTimeout">The time length in seconds before dictation recognizer session ends due to lack of audio input in case there was no audio heard in the current session.</param>
+        /// <param name="autoSilenceTimeout">The time length in seconds before dictation recognizer session ends due to lack of audio input.</param>
+        /// <param name="recordingTime">Length in seconds for the manager to listen.</param>
+        /// <returns></returns>
+        public static IEnumerator StartRecording(float initialSilenceTimeout = 5f, float autoSilenceTimeout = 20f, int recordingTime = 10)
+        {
+            if (IsListening) { yield break; }
```

Comment:
Hmm. This is a little tricky because Start/Stop recording in DictationInputManager is a coroutine that takes time to complete, but DictationRecordButton is a button that flips state instantaneously.

I have two ideas. Maybe you'll find one of these helpful or think of a better idea:

1) Have DictationInputManager consider it an error to call Stop when Start is running and hasn't yet completed. (Or to call Start when stop hasn't yet completed.) They could set a bool such as isTransitioning when they start their work, clear it when completed, and yield break when they are called if isTransitioning is already set. However, this would make the class more difficult to use. For example, DictationButton would have to track the return value of StartCoroutine and disable additional actions until it has completed.

2) Have DictationInputManager track its desired state with StartRecording/StopRecording just setting that state instantaneously. (StartRecording would set the parameters it wants to use in member variables. StopRecording would just save the fact that it wants recording to be off.) Then, you would need to write an update function that checks the current state, compares it to the desired state, and makes any calls needed to start/stop recording in response. This would be nice, but it seems complicated to write that logic to work correctly in all cases.